<a href="https://colab.research.google.com/github/techfreakydeepak/EDA_PLAYSTORE_APP_REVIEW_ANALYSIS/blob/main/EDA_PLAYSTORE_APP_REVIEW_ANALYSIS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
 #Let us begin with importing libraries

import pandas as pd # data manipulation
import numpy as np  # data manipulation 
import matplotlib.pyplot as plt # visualization tool
import seaborn as sns # visualization tool
import matplotlib

from datetime import datetime
from datetime import date

%matplotlib inline

In [2]:
# Mounting Google Drive 

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
file_path = "/content/drive/MyDrive/Colab Notebooks/eda /Play Store Data.csv"

In [4]:
reviews="/content/drive/MyDrive/Colab Notebooks/eda /User Reviews.csv"


In [5]:
df_apps = pd.read_csv(file_path)
df_reviews=pd.read_csv(reviews)

In [6]:
df_apps.head(10)

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   
5                         Paper flowers instructions  ART_AND_DESIGN     4.4   
6            Smoke Effect Photo Maker - Smoke Editor  ART_AND_DESIGN     3.8   
7                                   Infinite Painter  ART_AND_DESIGN     4.1   
8                               Garden Coloring Book  ART_AND_DESIGN     4.4   
9                      Kids Paint Free - Drawing Fun  ART_AND_DESIGN     4.7   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   
5     167  5.6M      50,000+  Free     0       Everyone   
6     178   19M      50,000+  Free     0       Everyone   
7   36815   29M   1,000,000+  Free     0       Everyone   
8   13791   33M   1,000,000+  Free     0       Everyone   
9     121  3.1M      10,000+  Free     0       Everyone   

                      Genres        Last Updated         Current Ver  \
0               Art & Design     January 7, 2018               1.0.0   
1  Art & Design;Pretend Play    January 15, 2018               2.0.0   
2               Art & Design      August 1, 2018               1.2.4   
3               Art & Design        June 8, 2018  Varies with device   
4    Art & Design;Creativity       June 20, 2018                 1.1   
5               Art & Design      March 26, 2017                   1   
6               Art & Design      April 26, 2018                 1.1   
7               Art & Design       June 14, 2018            6.1.61.1   
8               Art & Design  September 20, 2017               2.9.2   
9    Art & Design;Creativity        July 3, 2018                 2.8   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up  
5    2.3 and up  
6  4.0.3 and up  
7    4.2 and up  
8    3.0 and up  
9  4.0.3 and up

In [7]:
df_apps.tail(10)

App             Category  \
10831                       payermonstationnement.fr  MAPS_AND_NAVIGATION   
10832                                       FR Tides              WEATHER   
10833                                    Chemin (fr)  BOOKS_AND_REFERENCE   
10834                                  FR Calculator               FAMILY   
10835                                       FR Forms             BUSINESS   
10836                               Sya9a Maroc - FR               FAMILY   
10837               Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                         Parkinson Exercices FR              MEDICAL   
10839                  The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840  iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
10831     NaN      38                9.8M       5,000+  Free     0   
10832     3.8    1195                582k     100,000+  Free     0   
10833     4.8      44                619k       1,000+  Free     0   
10834     4.0       7                2.6M         500+  Free     0   
10835     NaN       0                9.6M          10+  Free     0   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating             Genres        Last Updated  \
10831       Everyone  Maps & Navigation       June 13, 2018   
10832       Everyone            Weather   February 16, 2014   
10833       Everyone  Books & Reference      March 23, 2014   
10834       Everyone          Education       June 18, 2017   
10835       Everyone           Business  September 29, 2016   
10836       Everyone          Education       July 25, 2017   
10837       Everyone          Education        July 6, 2018   
10838       Everyone            Medical    January 20, 2017   
10839     Mature 17+  Books & Reference    January 19, 2015   
10840       Everyone          Lifestyle       July 25, 2018   

              Current Ver         Android Ver  
10831           2.0.148.0          4.0 and up  
10832                   6          2.1 and up  
10833                 0.8          2.2 and up  
10834               1.0.0          4.1 and up  
10835               1.1.5          4.0 and up  
10836                1.48          4.1 and up  
10837                   1          4.1 and up  
10838                   1          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device

In [8]:
# print the columns of Two dataframes

print(f"The columns of Play Store Dataframe are \n \n {df_apps.columns}\n")
print(f"The columns of Play Store Dataframe are \n \n {df_reviews.columns}")


The columns of Play Store Dataframe are 
 
 Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

The columns of Play Store Dataframe are 
 
 Index(['App', 'Translated_Review', 'Sentiment', 'Sentiment_Polarity',
       'Sentiment_Subjectivity'],
      dtype='object')


In [9]:
print(f"The shape of Play Store dataframe is {df_apps.shape}")
print(f"The shape of User Review dataframe is {df_reviews.shape}")

The shape of Play Store dataframe is (10841, 13)
The shape of User Review dataframe is (64295, 5)


In [10]:
#
df_apps.columns = df_apps.columns.str.replace(" ","_")
df_apps.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content_Rating', 'Genres', 'Last_Updated', 'Current_Ver',
       'Android_Ver'],
      dtype='object')

Lets Explore The Play Store Dataset [ ]




In [11]:
# print the 10 head and 10 row of Play Store dataframe

concat_Playstore = pd.concat([df_apps.head(10),df_apps.tail(10)])
concat_Playstore

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
5                             Paper flowers instructions       ART_AND_DESIGN   
6                Smoke Effect Photo Maker - Smoke Editor       ART_AND_DESIGN   
7                                       Infinite Painter       ART_AND_DESIGN   
8                                   Garden Coloring Book       ART_AND_DESIGN   
9                          Kids Paint Free - Drawing Fun       ART_AND_DESIGN   
10831                           payermonstationnement.fr  MAPS_AND_NAVIGATION   
10832                                           FR Tides              WEATHER   
10833                                        Chemin (fr)  BOOKS_AND_REFERENCE   
10834                                      FR Calculator               FAMILY   
10835                                           FR Forms             BUSINESS   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
5         4.4     167                5.6M      50,000+  Free     0   
6         3.8     178                 19M      50,000+  Free     0   
7         4.1   36815                 29M   1,000,000+  Free     0   
8         4.4   13791                 33M   1,000,000+  Free     0   
9         4.7     121                3.1M      10,000+  Free     0   
10831     NaN      38                9.8M       5,000+  Free     0   
10832     3.8    1195                582k     100,000+  Free     0   
10833     4.8      44                619k       1,000+  Free     0   
10834     4.0       7                2.6M         500+  Free     0   
10835     NaN       0                9.6M          10+  Free     0   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content_Rating                     Genres        Last_Updated  \
0           Everyone               Art & Design     January 7, 2018   
1           Everyone  Art & Design;Pretend Play    January 15, 2018   
2           Everyone               Art & Design      August 1, 2018   
3               Teen               Art & Design        June 8, 2018   
4           Everyone    Art & Design;Creativity       June 20, 2018   
5           Everyone               Art & Design      March 26, 2017   
6           Everyone               Art & Design      April 26, 2018   
7           Everyone               Art & Design       June 14, 2018   
8           Everyone               Art & Design  September 20, 2017   
9           Everyone    Art & Design;Creativity        July 3, 2018   
10831       Everyone          Maps & Navigation       June 13, 2018   
10832       Everyone        

In [12]:
# print rows and column of play store data

print(f"There are total {df_apps.shape[0]} rows and {df_apps.shape[1]} columns in playstore dataframe.")



There are total 10841 rows and 13 columns in playstore dataframe.


In [13]:
df_apps.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content_Rating', 'Genres', 'Last_Updated', 'Current_Ver',
       'Android_Ver'],
      dtype='object')

In [14]:
df_apps.describe(include='all')

App Category       Rating Reviews                Size    Installs  \
count    10841    10841  9367.000000   10841               10841       10841   
unique    9660       34          NaN    6002                 462          22   
top     ROBLOX   FAMILY          NaN       0  Varies with device  1,000,000+   
freq         9     1972          NaN     596                1695        1579   
mean       NaN      NaN     4.193338     NaN                 NaN         NaN   
std        NaN      NaN     0.537431     NaN                 NaN         NaN   
min        NaN      NaN     1.000000     NaN                 NaN         NaN   
25%        NaN      NaN     4.000000     NaN                 NaN         NaN   
50%        NaN      NaN     4.300000     NaN                 NaN         NaN   
75%        NaN      NaN     4.500000     NaN                 NaN         NaN   
max        NaN      NaN    19.000000     NaN                 NaN         NaN   

         Type  Price Content_Rating Genres    Last_Updated  \
count   10840  10841          10840  10841           10841   
unique      3     93              6    120            1378   
top      Free      0       Everyone  Tools  August 3, 2018   
freq    10039  10040           8714    842             326   
mean      NaN    NaN            NaN    NaN             NaN   
std       NaN    NaN            NaN    NaN             NaN   
min       NaN    NaN            NaN    NaN             NaN   
25%       NaN    NaN            NaN    NaN             NaN   
50%       NaN    NaN            NaN    NaN             NaN   
75%       NaN    NaN            NaN    NaN             NaN   
max       NaN    NaN            NaN    NaN             NaN   

               Current_Ver Android_Ver  
count                10833       10838  
unique                2784          33  
top     Varies with device  4.1 and up  
freq                  1459        2451  
mean                   NaN         NaN  
std                    NaN         NaN  
min                    NaN         NaN  
25%                    NaN         NaN  
50%                    NaN         NaN  
75%                    NaN         NaN  
max                    NaN         NaN

In [15]:
def PlaystoreInfo():
  PSinfo_df = pd.DataFrame(index=df_apps.columns)
  PSinfo_df['Datatypes'] =  df_apps.dtypes
  PSinfo_df['Count of non-null values'] = df_apps.count()
  PSinfo_df['NaN values'] = df_apps.isnull().sum()
  PSinfo_df['% NaN Values'] = (PSinfo_df['NaN values']/len(df_apps)).round(4)*100 # or df_apps.isnull().mean()
  PSinfo_df['Unique_count'] = df_apps.nunique()
  return PSinfo_df
PlaystoreInfo()

Datatypes  Count of non-null values  NaN values  % NaN Values  \
App               object                     10841           0          0.00   
Category          object                     10841           0          0.00   
Rating           float64                      9367        1474         13.60   
Reviews           object                     10841           0          0.00   
Size              object                     10841           0          0.00   
Installs          object                     10841           0          0.00   
Type              object                     10840           1          0.01   
Price             object                     10841           0          0.00   
Content_Rating    object                     10840           1          0.01   
Genres            object                     10841           0          0.00   
Last_Updated      object                     10841           0          0.00   
Current_Ver       object                     10833           8          0.07   
Android_Ver       object                     10838           3          0.03   

                Unique_count  
App                     9660  
Category                  34  
Rating                    40  
Reviews                 6002  
Size                     462  
Installs                  22  
Type                       3  
Price                     93  
Content_Rating             6  
Genres                   120  
Last_Updated            1378  
Current_Ver             2784  
Android_Ver               33

In [16]:
# print the 10 head and 10 row of Play Store dataframe

concat_UserReview = pd.concat([df_reviews.head(10),df_reviews.tail(10)])
concat_UserReview

App  \
0            10 Best Foods for You   
1            10 Best Foods for You   
2            10 Best Foods for You   
3            10 Best Foods for You   
4            10 Best Foods for You   
5            10 Best Foods for You   
6            10 Best Foods for You   
7            10 Best Foods for You   
8            10 Best Foods for You   
9            10 Best Foods for You   
64285  Houzz Interior Design Ideas   
64286  Houzz Interior Design Ideas   
64287  Houzz Interior Design Ideas   
64288  Houzz Interior Design Ideas   
64289  Houzz Interior Design Ideas   
64290  Houzz Interior Design Ideas   
64291  Houzz Interior Design Ideas   
64292  Houzz Interior Design Ideas   
64293  Houzz Interior Design Ideas   
64294  Houzz Interior Design Ideas   

                                       Translated_Review Sentiment  \
0      I like eat delicious food. That's I'm cooking ...  Positive   
1        This help eating healthy exercise regular basis  Positive   
2                                                    NaN       NaN   
3             Works great especially going grocery store  Positive   
4                                           Best idea us  Positive   
5                                               Best way  Positive   
6                                                Amazing  Positive   
7                                                    NaN       NaN   
8                                   Looking forward app,   Neutral   
9                  It helpful site ! It help foods get !   Neutral   
64285                                                NaN       NaN   
64286                                                NaN       NaN   
64287                                                NaN       NaN   
64288                                                NaN       NaN   
64289                                                NaN       NaN   
64290                                                NaN       NaN   
64291                                                NaN       NaN   
64292                                                NaN       NaN   
64293                                                NaN       NaN   
64294                                                NaN       NaN   

       Sentiment_Polarity  Sentiment_Subjectivity  
0                    1.00                0.533333  
1                    0.25                0.288462  
2                     NaN                     NaN  
3                    0.40                0.875000  
4                    1.00                0.300000  
5                    1.00                0.300000  
6                    0.60                0.900000  
7                     NaN                     NaN  
8                    0.00                0.000000  
9                    0.00                0.000000  
64285                 NaN                     NaN  
64286                 NaN                     NaN  
64287                 NaN                     NaN  
64288                 NaN                     NaN  
64289                 NaN                     NaN  
64290                 NaN                     NaN  
64291                 NaN                     NaN  
64292                 NaN                     NaN  
64293                 NaN                     NaN  
64294                 NaN                     NaN

In [17]:
#print rows and column of play store data

print(f"There are total {df_reviews.shape[0]} rows and {df_reviews.shape[1]} columns in playstore dataframe.")

There are total 64295 rows and 5 columns in playstore dataframe.


In [18]:
df_reviews.columns

Index(['App', 'Translated_Review', 'Sentiment', 'Sentiment_Polarity',
       'Sentiment_Subjectivity'],
      dtype='object')

In [19]:
df_reviews.describe(include='all')

App Translated_Review Sentiment  Sentiment_Polarity  \
count                 64295             37427     37432        37432.000000   
unique                 1074             27994         3                 NaN   
top     Angry Birds Classic              Good  Positive                 NaN   
freq                    320               247     23998                 NaN   
mean                    NaN               NaN       NaN            0.182146   
std                     NaN               NaN       NaN            0.351301   
min                     NaN               NaN       NaN           -1.000000   
25%                     NaN               NaN       NaN            0.000000   
50%                     NaN               NaN       NaN            0.150000   
75%                     NaN               NaN       NaN            0.400000   
max                     NaN               NaN       NaN            1.000000   

        Sentiment_Subjectivity  
count             37432.000000  
unique                     NaN  
top                        NaN  
freq                       NaN  
mean                  0.492704  
std                   0.259949  
min                   0.000000  
25%                   0.357143  
50%                   0.514286  
75%                   0.650000  
max                   1.000000

In [20]:
def UserReviewInfo():
  URinfo_df = pd.DataFrame(index=df_reviews.columns)
  URinfo_df['Datatypes'] =  df_reviews.dtypes
  URinfo_df['Count of non-null values'] = df_reviews.count()
  URinfo_df['NaN values'] = df_reviews.isnull().sum()
  URinfo_df['% NaN Values'] = (URinfo_df['NaN values']/len(df_reviews)).round(5)*100 # or df_REVIEWS.isnull().mean()
  URinfo_df['Unique_count'] = df_reviews.nunique()
  return URinfo_df
UserReviewInfo()

Datatypes  Count of non-null values  NaN values  \
App                       object                     64295           0   
Translated_Review         object                     37427       26868   
Sentiment                 object                     37432       26863   
Sentiment_Polarity       float64                     37432       26863   
Sentiment_Subjectivity   float64                     37432       26863   

                        % NaN Values  Unique_count  
App                            0.000          1074  
Translated_Review             41.789         27994  
Sentiment                     41.781             3  
Sentiment_Polarity            41.781          5410  
Sentiment_Subjectivity        41.781          4474

Data Cleaning - Univariate & Bivariate Analysis


In [21]:
UserReviewInfo()

Datatypes  Count of non-null values  NaN values  \
App                       object                     64295           0   
Translated_Review         object                     37427       26868   
Sentiment                 object                     37432       26863   
Sentiment_Polarity       float64                     37432       26863   
Sentiment_Subjectivity   float64                     37432       26863   

                        % NaN Values  Unique_count  
App                            0.000          1074  
Translated_Review             41.789         27994  
Sentiment                     41.781             3  
Sentiment_Polarity            41.781          5410  
Sentiment_Subjectivity        41.781          4474

In [22]:
# lets create new user review df, as we going to drop all null values
# We have dropped the rows having all null values as it makes no sense in keeping them
new_df_reviews = df_reviews.dropna(subset=["Sentiment","Translated_Review","Sentiment_Polarity","Sentiment_Subjectivity"],how="all")

In [23]:
new_df_reviews.shape

(37432, 5)

In [24]:
new_df_reviews.isnull().sum()

App                       0
Translated_Review         5
Sentiment                 0
Sentiment_Polarity        0
Sentiment_Subjectivity    0
dtype: int64

In [25]:
new_df_reviews[new_df_reviews.Translated_Review.isnull()]

App Translated_Review Sentiment  \
268                                         11st               NaN   Neutral   
15048        Birds Sounds Ringtones & Wallpapers               NaN   Neutral   
22092             Calorie Counter - MyFitnessPal               NaN   Neutral   
31623                                  DC Comics               NaN   Neutral   
52500  Garden Photo Frames - Garden Photo Editor               NaN   Neutral   

       Sentiment_Polarity  Sentiment_Subjectivity  
268                   0.0                     0.0  
15048                 0.0                     0.0  
22092                 0.0                     0.0  
31623                 0.0                     0.0  
52500                 0.0                     0.0

In [26]:
new_df_reviews = new_df_reviews.dropna(subset= ["Translated_Review"],how = "all")
new_df_reviews.shape

(37427, 5)

In [27]:
def New_UserReviewInfo():
  URinfo_df = pd.DataFrame(index=new_df_reviews.columns)
  URinfo_df['Datatypes'] =  new_df_reviews.dtypes
  URinfo_df['Count of non-null values'] = new_df_reviews.count()
  URinfo_df['NaN values'] = new_df_reviews.isnull().sum()
  URinfo_df['% NaN Values'] = (URinfo_df['NaN values']/len(new_df_reviews)).round(5)*100 # or df_REVIEWS.isnull().mean()
  URinfo_df['Unique_count'] = new_df_reviews.nunique()
  return URinfo_df
New_UserReviewInfo()

Datatypes  Count of non-null values  NaN values  \
App                       object                     37427           0   
Translated_Review         object                     37427           0   
Sentiment                 object                     37427           0   
Sentiment_Polarity       float64                     37427           0   
Sentiment_Subjectivity   float64                     37427           0   

                        % NaN Values  Unique_count  
App                              0.0           865  
Translated_Review                0.0         27994  
Sentiment                        0.0             3  
Sentiment_Polarity               0.0          5410  
Sentiment_Subjectivity           0.0          4474

Missing Values in df_apps DF

In [28]:
# remove columns having more than 95% of the missing data

print(len(df_apps)*0.95) # 90% of data will be 9756
df_apps.isnull().sum()

# From the below output we can infer that no columns has 9756 NaN values 
# so we need not to Drop any Columns in this DataFrame.

10298.949999999999


App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content_Rating       1
Genres               0
Last_Updated         0
Current_Ver          8
Android_Ver          3
dtype: int64

In [29]:
# Lets check whether we have NAN values in all of the 5 columns which has NaN in each row. 

df_apps.dropna(subset=["Rating","Type","Content_Rating","Current_Ver","Android_Ver"],how="all").shape
df_apps.dropna(subset=["Rating","Type","Content_Rating","Current_Ver","Android_Ver"],how="all").shape

# since the shape of the output remains as original that means we dont have any rows where all of this values are NAN.

(10841, 13)

Rating Column



In [30]:
PlaystoreInfo()

Datatypes  Count of non-null values  NaN values  % NaN Values  \
App               object                     10841           0          0.00   
Category          object                     10841           0          0.00   
Rating           float64                      9367        1474         13.60   
Reviews           object                     10841           0          0.00   
Size              object                     10841           0          0.00   
Installs          object                     10841           0          0.00   
Type              object                     10840           1          0.01   
Price             object                     10841           0          0.00   
Content_Rating    object                     10840           1          0.01   
Genres            object                     10841           0          0.00   
Last_Updated      object                     10841           0          0.00   
Current_Ver       object                     10833           8          0.07   
Android_Ver       object                     10838           3          0.03   

                Unique_count  
App                     9660  
Category                  34  
Rating                    40  
Reviews                 6002  
Size                     462  
Installs                  22  
Type                       3  
Price                     93  
Content_Rating             6  
Genres                   120  
Last_Updated            1378  
Current_Ver             2784  
Android_Ver               33

In [ ]:
df_apps[df_apps["Rating"].isnull()].shape

Our Hypothesis:

We cannot simply drop these NaN values of Rating Column.

So in order to get sentiment or translated review info for the apps from the user_review DF so that we can fill these NaN values based a logic based derivation.

Thus, we need to merge these two DF's so that we get Common Apps between between them and can compare.

Lets Merge Two DataFrames based on common "App" Column.

In [31]:
# Lets Merge two DataFrame to check the common apps 
df_apps_copy = df_apps.copy()
new_df_reviews_copy = new_df_reviews.copy()

In [32]:
merged_apps_reviews = pd.merge(df_apps_copy, new_df_reviews_copy, how='inner', left_on='App', right_on='App')

In [33]:
def Mergedinfo():
    temp = pd.DataFrame(index=merged_apps_reviews.columns)
    temp['data_type'] = merged_apps_reviews.dtypes
    temp["count of non null values"] = merged_apps_reviews.count()
    temp['NaN values'] = merged_apps_reviews.isnull().sum()
    temp['% NaN values'] = (temp['NaN values']/len(merged_apps_reviews)).round(5)*100 # or merged_apps_reviews.isnull().mean()
    temp['unique_count'] = merged_apps_reviews.nunique()
    return temp 
Mergedinfo()

data_type  count of non null values  NaN values  \
App                       object                     72605           0   
Category                  object                     72605           0   
Rating                   float64                     72566          39   
Reviews                   object                     72605           0   
Size                      object                     72605           0   
Installs                  object                     72605           0   
Type                      object                     72605           0   
Price                     object                     72605           0   
Content_Rating            object                     72605           0   
Genres                    object                     72605           0   
Last_Updated              object                     72605           0   
Current_Ver               object                     72605           0   
Android_Ver               object                     72605           0   
Translated_Review         object                     72605           0   
Sentiment                 object                     72605           0   
Sentiment_Polarity       float64                     72605           0   
Sentiment_Subjectivity   float64                     72605           0   

                        % NaN values  unique_count  
App                            0.000           816  
Category                       0.000            33  
Rating                         0.054            22  
Reviews                        0.000          1068  
Size                           0.000           168  
Installs                       0.000            12  
Type                           0.000             2  
Price                          0.000             9  
Content_Rating                 0.000             5  
Genres                         0.000            68  
Last_Updated                   0.000           251  
Current_Ver                    0.000           504  
Android_Ver                    0.000            22  
Translated_Review              0.000         26682  
Sentiment                      0.000             3  
Sentiment_Polarity             0.000          5295  
Sentiment_Subjectivity         0.000          4382

Form Above we can say that their are 39 value are missing.

In [ ]:
print(df_apps.App.nunique())
print(new_df_reviews.App.nunique())
print(Mergedinfo()["unique_count"]["App"])

From the Above output we can see that we have 9660 unique Apps in Playstore DF.865 unique Apps in UserReview DF.

From the Merged DF we can infer that, we have only 816 Common Apps between Two DF's.

In [34]:
merged_apps_reviews[merged_apps_reviews.Rating.isnull()] 

App Category  Rating Reviews  Size Installs  Type Price  \
47874  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47875  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47876  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47877  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47878  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47879  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47880  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47881  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47882  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47883  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47884  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47885  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47886  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47887  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47888  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47889  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47890  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47891  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47892  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47893  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47894  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47895  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47896  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47897  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47898  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47899  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47900  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47901  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47902  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47903  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47904  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47905  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47906  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47907  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47908  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47909  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47910  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47911  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
47912  Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   

      Content_Rating   Genres    Last_Updated Current_Ver Android_Ver  \
47874       Everyone  Medical  March 14, 2015           1  2.3 and up   
47875       Everyone  Medical  March 14, 2015           1  2.3 and up   
47876       Everyone  Medical  March 14, 2015           1  2.3 and up   
47877       Everyone  Medical  March 14, 2015           1  2.3 and up   
47878       Everyone  Medical  March 14, 2015           1  2.3 and up   
47879       Everyone  Medical  March 14, 2015           1  2.3 and up   
47880       Everyone  Medical  March 14, 2015           1  2.3 and up   
47881       Everyone  Medical  March 14, 2015           1  2.3 and up   
47882       Everyone  Medical  March 14, 2015           1  2.3 and up   
47883       Everyone  Medical  March 14, 2015           1  2.3 and up   
47884       Everyone  Medical  March 14, 2015           1  2.3 and up   
47885       Everyone  Medical  March 14, 2015           1  2.

In [ ]:
l = len(merged_apps_reviews[merged_apps_reviews.Rating.isnull()]) # 39
l1 = len(merged_apps_reviews[(merged_apps_reviews.Rating.isnull()) & (merged_apps_reviews["Sentiment"].str.contains("Positive"))]) # 31
l2 = len(merged_apps_reviews[(merged_apps_reviews.Rating.isnull()) & (merged_apps_reviews["Sentiment"].str.contains("Negative"))]) # 4
l3 = len(merged_apps_reviews[(merged_apps_reviews.Rating.isnull()) & (merged_apps_reviews["Sentiment"].str.contains("Neutral"))]) # 4

positive_per = round((l1/l)*100,2)
negative_per = round((l2/l)*100,2)
Neutral_per = round((l3/l)*100,2)

print(negative_per,positive_per,Neutral_per)

In [ ]:
merged_apps_reviews.Sentiment.unique()

From the above output we can infer that the App for which Rating are NaN is "Blood Pressure" and rest of the values are also same out of which Sentiment column has - 10% of negative and 10% are neutral 80% are positive values for the App "Blood Pressure"

Let us now check the rating NaN values in df_apps DF

In [35]:
df_apps[df_apps["Rating"].isnull() & df_apps["Category"].str.contains("MEDICAL") & df_apps["App"].str.contains("Blood Pressure")]

App Category  Rating Reviews  Size Installs  Type Price  \
2356  Free Blood Pressure  MEDICAL     NaN       7  5.7M   5,000+  Free     0   
2502  Free Blood Pressure  MEDICAL     NaN       7  5.7M   5,000+  Free     0   
2513       Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
6598  Free Blood Pressure  MEDICAL     NaN       7  5.7M   5,000+  Free     0   

     Content_Rating   Genres      Last_Updated Current_Ver   Android_Ver  
2356       Everyone  Medical  October 13, 2016       3.0.0  4.0.3 and up  
2502       Everyone  Medical  October 13, 2016       3.0.0  4.0.3 and up  
2513       Everyone  Medical    March 14, 2015           1    2.3 and up  
6598       Everyone  Medical  October 13, 2016       3.0.0  4.0.3 and up

In [ ]:
mode_value = df_apps.Rating.mode()[0]
mode_value

Since in the Merged DF , the Sentiment column has - 10% of negative and 10% are neutral 80% are positive values for the App "Blood Pressure"


Hence since we have 80% of positive reviews for "Blood Pressure App" we can fill the NaN of rating with mode value "4.4"

In [ ]:
df_apps.loc[2515] = df_apps.loc[2515].fillna(mode_value)

In [ ]:
df_apps.iloc[2515]

In [36]:
# This is just for confirmation after filling one NaN Rating. 

print(f'The shape of df_apps DF having NaN Ratings are : {df_apps[df_apps["Rating"].isnull()].shape}')
df_apps[df_apps["Rating"].isnull() & df_apps["Category"].str.contains("MEDICAL") & df_apps["App"].str.contains("Blood Pressure")]

The shape of df_apps DF having NaN Ratings are : (1474, 13)


App Category  Rating Reviews  Size Installs  Type Price  \
2356  Free Blood Pressure  MEDICAL     NaN       7  5.7M   5,000+  Free     0   
2502  Free Blood Pressure  MEDICAL     NaN       7  5.7M   5,000+  Free     0   
2513       Blood Pressure  MEDICAL     NaN      10  2.4M   1,000+  Free     0   
6598  Free Blood Pressure  MEDICAL     NaN       7  5.7M   5,000+  Free     0   

     Content_Rating   Genres      Last_Updated Current_Ver   Android_Ver  
2356       Everyone  Medical  October 13, 2016       3.0.0  4.0.3 and up  
2502       Everyone  Medical  October 13, 2016       3.0.0  4.0.3 and up  
2513       Everyone  Medical    March 14, 2015           1    2.3 and up  
6598       Everyone  Medical  October 13, 2016       3.0.0  4.0.3 and up

Findings

Hypothesis 1: We don't have infomation of sentiment of the Apps to fill the NaN values for the remaining apps from the merged DF.

Hypothesis 2: Let us check whether do we have rating of Apps within the PlayStore DF.

In [37]:
# Unique Categories in the filtered DF where the ratings are null.

df_apps[df_apps["Rating"].isnull()]["Category"].unique()

array(['ART_AND_DESIGN', 'BEAUTY', 'BOOKS_AND_REFERENCE', 'BUSINESS',
       'COMICS', 'DATING', 'EDUCATION', 'EVENTS', 'FOOD_AND_DRINK',
       'HOUSE_AND_HOME', 'LIBRARIES_AND_DEMO', 'FAMILY', 'MEDICAL',
       'TOOLS', 'PARENTING', 'VIDEO_PLAYERS', 'PERSONALIZATION', 'GAME',
       'PHOTOGRAPHY', 'SOCIAL', 'SPORTS', 'COMMUNICATION', 'PRODUCTIVITY',
       'AUTO_AND_VEHICLES', 'FINANCE', 'LIFESTYLE', 'TRAVEL_AND_LOCAL',
       'SHOPPING', 'HEALTH_AND_FITNESS', 'WEATHER', 'NEWS_AND_MAGAZINES',
       'MAPS_AND_NAVIGATION'], dtype=object)

In [ ]:
df_apps.groupby(["Category"])["App"].nunique()

From the above we can see that number of unique Apps found in each category.

Note: Ignore the 1.9 category name as of now.

In [ ]:
df_apps[(df_apps["Category"]== "AUTO_AND_VEHICLES") & (df_apps["App"].str.contains("FORD V SERIES CALC - NO LIMIT"))]

If we compare the count of Apps and Unique count of Apps above we can get an idea that the count and unique count are almost equal in each category.

Findings

Hypothesis 2: From this we can see that even with this hypothesis we dont get information needed to fill NaN of the Rating.

Since both Hypothesis did not get us the information about Filling NaN values of Rating. 

We shall proceed with dropping these from the df_apps DF.

In [ ]:
new_df_apps = df_apps.dropna(subset = ["Rating"],how= "all")

Let us check for any noises or irrelevant data in any of the Columns/Rows and perform the following on our df_apps DF.

1.Reviews should be numeric type.

2.Size should be numeric type. In the given dataset it has 'M' and 'k' strings present in it. We need to convert these into numbers.

3.Installs should be numeric type. It has '+' & ',' characters present in it, which needs to be removed.

4.Price should be numeric type. It has '$' character present in it, which needs to be removed.


6.Android_Ver should be numeric type.

7.Junk Apps should be filtered out from the dataframe.